### DISCUSSION TITLE
Is there any option like middleware to extend react-query

### DISCUSSION THREAD

**Discussion by krishnasaga** [TanStack/query#5877] [Q&A]
Having middleware like extendable feature allows to create more higher-level abstractions like convention-based cache keys.

useSWR has middleware feature.
https://swr.vercel.app/docs/middleware

**Comment by TkDodo** [comment]
what would be a use-case? all the examples listed on that docs page can be achieved differently with built-in solutions

**Reply by krishnasaga** [reply]
`what would be a use-case?`
It is to leave the room for unknowns rather assuming covered all use cases in the world. In reality, there can be always a use case outside the examples listed in the docs page. `useSWR` and redux based solutions has that option.  Which makes it is possible to have higher level constraints specific to the project needs especially for larger projects like name spacing local storage, instrumenting performance bottlenecks and Erros etc. Thinking to choose between react-query and useSWR

**Reply by TkDodo** [reply]
on the flip side, middlewares are a complicated concept to understand. @tannerlinsley said a while back that this is not what we want for this lib:

- https://twitter.com/tannerlinsley/status/1433154450600853505

**Reply by devth** [reply]
I'm thinking middleware would be nice for handling errors from server actions (where it's still a 200 so `onError` isn't triggered). Maybe there's a better way though.
